knowhere/index/vector_index/helpers/RangeUtil.h

```diff
@@ -0,0 +1,161 @@
+                             const float high_bound,
+                             const int64_t i_size,
+                             const float* i_distances,
+                             const int64_t* i_labels,
+                             const int64_t o_size,
+                             float* o_distances,
+                             int64_t* o_labels,
+                             const bool is_ip,
+                             const faiss::BitsetView bitset) {
+    size_t num = 0;
+    for (size_t i = 0; i < i_size; i++) {
+        auto dis = i_distances[i];
+        auto id = i_labels[i];
+        assert(bitset.empty() || !bitset.test(id));
+        assert((is_ip && dis > low_bound) || (!is_ip && dis < high_bound));
+        if (distance_in_range(dis, low_bound, high_bound, is_ip)) {
+            o_labels[num] = id;
+            o_distances[num] = dis;
+            num++;
+        }
+    }
+    assert(num == o_size);
+}
+
+inline void
+GetRangeSearchResult(const faiss::RangeSearchResult& res,
+                     const faiss::MetricType metric_type,
+                     const int64_t nq,
+                     const float low_bound,
+                     const float high_bound,
+                     float*& distances,
+                     int64_t*& labels,
+                     size_t*& lims,
+                     const faiss::BitsetView bitset) {
+    TimeRecorder rc("GetRangeSearchResult");
+
+    bool is_ip = (metric_type == faiss::METRIC_INNER_PRODUCT);
+    lims = new size_t[nq + 1];
+
+    lims[0] = 0;
+    for (int64_t i = 0; i < nq; i++) {
+        int64_t valid = 0;
+        for (size_t j = res.lims[i]; j < res.lims[i+1]; j++) {
+            if (distance_in_range(res.distances[j], low_bound, high_bound, is_ip)) {
+                valid++;
+            }
+        }
+        lims[i + 1] = lims[i] + valid;
+    }
+
+    size_t total_valid = lims[nq];
+    LOG_KNOWHERE_DEBUG_ << "Range search metric type: " << metric_type << ", low_bound " << low_bound << ", high_bound "
+                        << high_bound << ", total result num: " << total_valid;
+
+    distances = new float[total_valid];
+    labels = new int64_t[total_valid];
+
+#pragma omp parallel for
+    for (auto i = 0; i < nq; i++) {
+        GetRangeSearchResultForOneNq(low_bound,
+                                     high_bound,
+                                     res.lims[i + 1] - res.lims[i],
+                                     res.distances + res.lims[i],
+                                     res.labels + res.lims[i],
+                                     lims[i + 1] - lims[i],
+                                     distances + lims[i],
+                                     labels + lims[i],
+                                     is_ip,
+                                     bitset);
+    }
+
+    rc.ElapseFromBegin("done");
+}
+
+// for HNSW
+inline void
+GetRangeSearchResult(const std::vector<std::vector<float>>& result_distances,
```

Comment:
This sounds weird to me:
1. We do range check in multiple places: indexes, here,  and in `GetRangeSearchResult`. 
2. The post processing init another omp block which is less efficient. We can put then in a same thread where the processing happen.

Reply:
Have optimized the post process for HNSW and DiskANN.